### PR TITLE
XDSM writer (pyXDSM and XDSMjs)

### PR DESCRIPTION
**Pull request by onodip**
Added an XDSM writer based on the pyXDSM package (so it generates TeX file and PDF).  
Added tests (skipped, if pyXDSm is not installed, so there will be no issues hopefully with the dependencies).

There is an issue in the pyXDSM package with filenames on Windows. The workaround added in the docstring of the `XDSMWriter`

**Known issues:**
* The current implementation is not attempting to make a perfect XDSM, just tries to automate some part of the process, and might need manual editing.

### CODE REVIEW

**Comment by onodip** [comment]
The output from the second test looks like this:

xdsm

**Comment by JustinSGray** [comment]
as an aside, i've been thinking about integrating XDSMjs into the model_viewer. @onodip any interest in taking that on? 
https://github.com/OneraHub/XDSMjs

**Comment by onodip** [comment]
@JustinSGray I like the idea. And it should not be difficult to generate the JSON file for it. Now I am a bit overwhelmed with my thesis, but if it is not something very urgent, I can do it.

**Comment by JustinSGray** [comment]
Its not urgent, and the current pyXDSM integration is a good start.

**Comment by onodip** [comment]
@JustinSGray I made a very quick proof-of-concept, a lot of things are missing yet. It reuses the `write_xdsm()` function from the PR, with another argument for the output format, and I wrote a class similar to the `XDSMWriter` to make the JSON file.

xdsmjs0

**Comment by onodip** [comment]
Added the preliminary version of the XDSMjs writer to the PR. For now it only generates the JSON file (not possible to open the HTML from OpenMDAO yet), and it is quite basic. 
@robfalck with the variable naming can you take into account the restrictions (e.g. forbidden characters, etc) for both pyXDSM and XDSMjs?

**Comment by robfalck** [comment]
> Added the preliminary version of the XDSMjs writer to the PR. For now it only generates the JSON file (not possible to open the HTML from OpenMDAO yet), and it is quite basic.
> @robfalck with the variable naming can you take into account the restrictions (e.g. forbidden characters, etc) for both pyXDSM and XDSMjs?

I think the way you've implemented it will suffice.  I added a PR to your original branch that accommodates my use-cases (only made an XDSM of a submodel, and make recursion into subgroups optional).

You may want to move all of your helper functions out of _write_xdsm and indicate that they aren't functions the user should use by prepending them with an underscore.  That would be a bit more in line with the style of the rest of OpenMDAO.

**Comment by swryan** [comment]
> No clue about the source of this error, it is not raised with `testflo `on my machine:
> 
> `AttributeError: 'Test' object has no attribute 'isolated'`

this looks like a bug in `testflo` introduced by a recent change..   will fix.

**Comment by onodip** [comment]
@robfalck thanks for the PR, I merged it, and changed some nested functions to private functions.

**Comment by swryan** [comment]
> > No clue about the source of this error, it is not raised with `testflo `on my machine:
> > `AttributeError: 'Test' object has no attribute 'isolated'`
> 
> this looks like a bug in `testflo` introduced by a recent change.. will fix.

This is the error that is being masked by the `testflo` bug... I can see this when testing on a fork of your branch:

> /home/travis/miniconda/envs/PY3.6/lib/python3.6/site-packages/openmdao/devtools/tests/test_xdsm_viewer.py ... FAIL (00:00:0.00, 0 MB)
> Traceback (most recent call last):
>   File "/home/travis/miniconda/envs/PY3.6/lib/python3.6/site-packages/openmdao/devtools/xdsm_writer.py", line 12, in <module>
>     from pyxdsm.XDSM import XDSM
> ModuleNotFoundError: No module named 'pyxdsm'
> During handling of the above exception, another exception occurred:
> Traceback (most recent call last):
>   File "/home/travis/miniconda/envs/PY3.6/lib/python3.6/site-packages/testflo/discover.py", line 129, in _module_iter
>     fname, mod = get_module(filename)
>   File "/home/travis/miniconda/envs/PY3.6/lib/python3.6/site-packages/testflo/util.py", line 314, in get_module
>     __import__(modpath)
>   File "/home/travis/miniconda/envs/PY3.6/lib/python3.6/site-packages/openmdao/devtools/tests/test_xdsm_viewer.py", line 7, in <module>
>     from openmdao.devtools.xdsm_writer import write_xdsm
>   File "/home/travis/miniconda/envs/PY3.6/lib/python3.6/site-packages/openmdao/devtools/xdsm_writer.py", line 16, in <module>
>     raise RuntimeError(msg)
> RuntimeError: The pyxdsm package should be installed. You can download the package from https://github.com/mdolab/pyXDSM

**Comment by onodip** [comment]
> > > No clue about the source of this error, it is not raised with `testflo `on my machine:
> > > `AttributeError: 'Test' object has no attribute 'isolated'`
> > 
> > 
> > this looks like a bug in `testflo` introduced by a recent change.. will fix.
> 
> This is the error that is being masked by the `testflo` bug... I can see this when testing on a fork of your branch:
> 
> > /home/travis/miniconda/envs/PY3.6/lib/python3.6/site-packages/openmdao/devtools/tests/test_xdsm_viewer.py ... FAIL (00:00:0.00, 0 MB)
> > Traceback (most recent call last):
> > File "/home/travis/miniconda/envs/PY3.6/lib/python3.6/site-packages/openmdao/devtools/xdsm_writer.py", line 12, in 
> > from pyxdsm.XDSM import XDSM
> > ModuleNotFoundError: No module named 'pyxdsm'
> > During handling of the above exception, another exception occurred:
> > Traceback (most recent call last):
> > File "/home/travis/miniconda/envs/PY3.6/lib/python3.6/site-packages/testflo/discover.py", line 129, in _module_iter
> > fname, mod = get_module(filename)
> > File "/home/travis/miniconda/envs/PY3.6/lib/python3.6/site-packages/testflo/util.py", line 314, in get_module
> > **import**(modpath)
> > File "/home/travis/miniconda/envs/PY3.6/lib/python3.6/site-packages/openmdao/devtools/tests/test_xdsm_viewer.py", line 7, in 
> > from openmdao.devtools.xdsm_writer import write_xdsm
> > File "/home/travis/miniconda/envs/PY3.6/lib/python3.6/site-packages/openmdao/devtools/xdsm_writer.py", line 16, in 
> > raise RuntimeError(msg)
> > RuntimeError: The pyxdsm package should be installed. You can download the package from https://github.com/mdolab/pyXDSM

Okay, basically I am trying to suppress an ImportError from a package not listed in the requirements. Any better way to do this?

**Comment by swryan** [comment]
Typically we have been catching the import failure at the top of  the test script and using that as a skip condition  (see `test_bsplines_comp.py` for example).  Following that pattern you could only import `write_xdsm` if the `XDSM` import was successful... e.g.

> try:
>     from pyxdsm.XDSM import XDSM
>     from openmdao.devtools.xdsm_writer import write_xdsm
> except ImportError:
>     XDSM = None
> 
> 
> @unittest.skipUnless(XDSM, "XDSM is required.")
> class TestXDSMViewer(unittest.TestCase):

That's just a suggestion based on how we have been doing it...

**Comment by onodip** [comment]
> Typically we have been catching the import failure at the top of the test script and using that as a skip condition (see `test_bsplines_comp.py` for example). Following that pattern you could only import `write_xdsm` if the `XDSM` import was successful... e.g.
> 
> ```
> try:
>     from pyxdsm.XDSM import XDSM
>     from openmdao.devtools.xdsm_writer import write_xdsm
> except ImportError:
>     XDSM = None
> 
> @unittest.skipUnless(XDSM, "XDSM is required.")
> class TestXDSMViewer(unittest.TestCase):
> ```
> That's just a suggestion based on how we have been doing it...

Thank you @swryan, I included your exception handling, and it works fine.

**Comment by onodip** [comment]
Now the output file can be opened with the "show_browser" option similarly as with `view_model()` for both _pyXDSM_ and _XDSMjs_.

For _XDSMjs_ this is not very nice, because the JSON file have to be copied to the program's folder and renamed to "xdsm.json". For now the user have to give the XDSMjs folder as an argument, I could not come with anything better so far.

**Comment by JustinSGray** [comment]
Can the html file be modified in such a way as to hard-code the json data directly into the file? This is how its done for the N2

**Comment by onodip** [comment]
> Can the html file be modified in such a way as to hard-code the json data directly into the file? This is how its done for the N2

The goal would be to make the HTML file standalone? In this case the javascript codes also have to be copied into it besides the JSON data. I can look into it.

**Comment by JustinSGray** [comment]
yes, a stand alone html file. That freezes the xdsm.js version, but its worth it to have a portable file that can work simply.

**Comment by relf** [comment]
Hi! XDSMjs author here, happy to see XDSMjs about to be embedded in OpenMDAO (thanks @onodip). I've just tagged a [0.5.0 version](https://github.com/OneraHub/XDSMjs/releases/tag/v0.5.0) to celebrate the event. :tada:

**Comment by onodip** [comment]
@relf thanks for the new version! I've made some work on embedding everything into the HTML file. So far it is in a [separate branch](https://github.com/onodip/OpenMDAO/tree/xdsmjs). I embedded the CSS and JavaScript files, and the plan is to make it work both by referencing a JSON file or embedding the data and make the HTML completely standalone (I have problems with the latter approach so far).

**Comment by relf** [comment]
Regarding JSON data embedding, I've just implemented in XDSMjs the ability to specify JSON data through data attributes (see OneraHub/XDSMjs#4). I think it is the way to go to complete your implementation.

File: openmdao/devtools/xdsm_viewer/html_writer.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+    # grab the libraries
+    scripts = ''
+    script_names = {'animation', 'controls', 'graph', 'labelizer', 'selectable', 'xdsm',
+                    'xdsm-factory'}
+
+    with open(os.path.join(build_dir, "xdsm.bundle.js"), "r") as f:
+        code = f.read()
+        xdsm_bundle = _write_script(code, {'type': 'text/javascript'})
+    # grab the scripts
```
Comment (relf) [inline openmdao/devtools/xdsm_viewer/html_writer.py]: You don't need to grab the js files from XDSMjs src directory, all you need is the xdsm.bundle.js file which is actually the output of the XDSMjs build process which uses those js source files as inputs.

Comment (onodip) [reply openmdao/devtools/xdsm_viewer/html_writer.py]: Thanks for the explanation, corrected.

**Comment by onodip** [comment]
On my side I don't plan any further changes on this PR, I will be waiting for the review.

**Comment by Kenneth-T-Moore** [comment]
Hey, Péter

I tried out your writer (on one of the harder models). Very nice.  There are just a couple of things I noticed.

1. If you have an underscore in a variable or component name, the LaTeX, it is replaced by a space in most blocks, which looks kind of strange. I would prefer underscores there. In the output blocks (white parallelograms), they aren't escaped properly, so the character after the underscore becomes a subscript.

2. If you have a lot of long variable names in the intermediate connections, the blocks become very long. I like how the variables are stacked vertically in the connections that come from the green Driver block, so it might be nice to stack inter-component connections that way too, once they are longer than a certain length.

3. Found some missing connections in a dymos model. Dymos does things a little differently than a conventional model, but I can see the connections when I run view_model, so they are real. I am going to try to reduce this to something that doesn't contain Dymos to see if I can isolate the bug.

**Comment by Kenneth-T-Moore** [comment]
Ok, found a simple model to reproduce the bug!

https://gist.github.com/Kenneth-T-Moore/7f499bff2d52b174ae0de2dd37449b37

What I think is happening is the xdsm writer is getting confused when there are multiple instances of a group, so that there are components and connections with the same local names (though they have unique global names).  The connection info is all being packed into the last group's connection.  Compare the output of the model viewer with the output of the xdsm writer in the example above, and you'll see.  Probably an easy fix.

**Comment by onodip** [comment]
@Kenneth-T-Moore thanks for the feedback and for the example. 

I made the following changes:
1. Now the inputs & outputs are formatted the same way as the components
2. The variable names in the connections can be stacked vertically, horizontally, split to multiple lines if they are long, or show only the some of them, if they are long.  I think if anything fancier is needed, the pyXDSM package has to be modified a bit.
3. It was indeed because of the non-unique relative var names. Now I am using absolute names internally for the component names and connections. For the labels I am using the absolute names only for the non-unique components (maybe should be an option to switch between this, rel. names only and abs. names only)

**Some examples:**

Non-unique variable names. The path is included in the names.

image

Cut lines where line>15 characters

image

Variables stacked, so each line<=15 characters
image

**Comment by Kenneth-T-Moore** [comment]
Thanks, the fixes all look great.

**Comment by onodip** [comment]
...and the same with XDSMjs:

image

**Comment by onodip** [comment]
Added process connections (the thin black lines) and initial design var blocks and fixed some formatting issues.

I think now I am done for now.

image

**Comment by JustinSGray** [comment]
Thanks for the awesome contribution. Super valuable!

**Comment by JustinSGray** [comment]
@onodip it looks like there are some files missing from the repo. The XDSM works with pyxdsm, but when I tried to generate one with XDSMjs I got some errors realted to missing files: 

```FileNotFoundError: [Errno 2] No such file or directory: '/Users/jsgray/work/packages/OpenMDAO/openmdao/devtools/xdsm_viewer/XDSMjs/build/xdsm.bundle.js
FileNotFoundError: [Errno 2] No such file or directory: '/Users/jsgray/work/packages/OpenMDAO/openmdao/devtools/xdsm_viewer/XDSMjs/fontello.css
FileNotFoundError: [Errno 2] No such file or directory: '/Users/jsgray/work/packages/OpenMDAO/openmdao/devtools/xdsm_viewer/XDSMjs/xdsm.css
FileNotFoundError: [Errno 2] No such file or directory: '/Users/jsgray/work/packages/OpenMDAO/openmdao/devtools/xdsm_viewer/visualization/index.html
```

could you submit a new PR with the missing files added to the repo? 
Thanks,

**Comment by relf** [comment]
Maybe a problem with the XDSMjs directory as a git submodule. I think there is a missing .gitmodules in the root directory.

**Comment by onodip** [comment]
Oops, trying to fix in #842.

**Pull request merged**